// small-google-calendar/dist/src/components/Calendar/types.js
export var LocalDay;
(function (LocalDay) {
    LocalDay[LocalDay["Sunday"] = 0] = "Sunday";
    LocalDay[LocalDay["Monday"] = 1] = "Monday";
    LocalDay[LocalDay["Tuesday"] = 2] = "Tuesday";
    LocalDay[LocalDay["Wednesday"] = 3] = "Wednesday";
    LocalDay[LocalDay["Thursday"] = 4] = "Thursday";
    LocalDay[LocalDay["Friday"] = 5] = "Friday";
    LocalDay[LocalDay["Saturday"] = 6] = "Saturday";
})(LocalDay || (LocalDay = {}));
export var Month;
(function (Month) {
    Month[Month["January"] = 1] = "January";
    Month[Month["February"] = 2] = "February";
    Month[Month["March"] = 3] = "March";
    Month[Month["April"] = 4] = "April";
    Month[Month["May"] = 5] = "May";
    Month[Month["June"] = 6] = "June";
    Month[Month["July"] = 7] = "July";
    Month[Month["August"] = 8] = "August";
    Month[Month["September"] = 9] = "September";
    Month[Month["October"] = 10] = "October";
    Month[Month["November"] = 11] = "November";
    Month[Month["December"] = 12] = "December";
})(Month || (Month = {}));
